[SYCL] Add group local memory call lowering pass

mlychkov opened the pull request:
The following call is included into device code for work group
by SYCL headers to declare static local memory allocation in SYCL kernel:
'__attribute__((opencl_local)) uint8_t *Ptr = __sycl_allocateLocalMemory(sizeof(T), alignof(T))'.
Then during compilation it should be transformed somewhere to
equivalent of the following code:
'__local alignas(alignof(T)) uint8_t Mem[sizeof(T)];
 __local uint8_t *Ptr = &Mem;'

Add a pass that handles creation of fixed-size allocations in local
address space at the kernel scope.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

kbobrovs reviewed (changes requested):
- Why is this needed, what problem does this patch solve?
- Why this is part of post-link?
- Please add tests

mlychkov commented:
> * Why is this needed, what problem does this patch solve?
> * Why this is part of post-link?
> * Please add tests

* Updated commit message.
* Moved pass to clang FE.
* Added test.

elizabethandrews commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -948,6 +948,10 @@ void EmitAssemblyHelper::EmitAssembly(BackendAction Action,
       LangOpts.EnableDAEInSpirKernels)
     PerModulePasses.add(createDeadArgEliminationSYCLPass());
 
+  // This pass should be always called for SYCL device code.
```
Please add a frontend test as well.

mlychkov replied:
Done

bader commented on llvm/test/Transforms/LowerWGLocalMemoryCall/group_local_memory.ll:
```diff
@@ -0,0 +1,286 @@
+; RUN: opt -S -sycllowerwglocalmemorycall < %s | FileCheck --check-prefixes=CHECK,CHECK-CALL %s
```
Please, reduce the test case by removing unnecessary instructions, metadata, etc.

BTW, why do we need `CHECK-CALL` prefix?

mlychkov replied:
It is used to check that there is no calls of __sycl_allocateLocalMemory left in IR after the pass.

bader commented on llvm/test/Transforms/LowerWGLocalMemoryCall/group_local_memory.ll:
```diff
@@ -0,0 +1,286 @@
+
+if.then.i23.i:                                    ; preds = %entry
+  %14 = bitcast i8 addrspace(3)* %call.i22.i to i32 addrspace(3)*
+  ; CHECK-COUNT-33: i8 addrspace(3)* getelementptr inbounds ([128 x i8], [128 x i8] addrspace(3)* [[WGLOCALMEM_1]], i32 0, i32 0)
```
Can we reduce the number of instructions? What is the reason to have 33 similar instructions?

mlychkov replied:
Done

bader commented on llvm/lib/Transforms/Utils/LowerWGLocalMemoryCall.cpp:
```diff
@@ -0,0 +1,114 @@
+      StringRef Name = Callee->getName();
+      if (Name == SYCL_ALLOCLOCALMEM_CALL)
+        ToReplace.push_back(CI);
+    }
```
It would be useful to add an assertion to verify that intrinsic allocates memory only in a kernel scope.

kbobrovs replied:
+1.
Also, can this occur inside if? Inside a loop? If not - assertions could be added too.

mlychkov replied:
@kbobrovs Sorry, I didn't get this comment. Could you please clarify it?

kbobrovs replied:
@mlychkov, is it legal for __sycl_allocateLocalMemory to happen:
- inside a loop
- inside a block not dominating exit
?

mlychkov replied:
@Pennycook Could you please help to answer this question?

bader replied:
No, it's not allowed, but LLVM pass is not the best place to diagnose these cases. Error checking should be done in the front-end compiler.

> It would be useful to add an assertion to verify that intrinsic allocates memory only in a kernel scope.

It turns out that even check like this is difficult to implement in LLVM pass with the current design. We would like to allow `__sycl_allocateLocalMemory` invocations in lambda call operator's scope, which is technically not SPIR-V kernel scope.

kbobrovs replied:
> No, it's not allowed

Right, that's what I expected. Checking validity should somehow be addressed. Plus one more case - non-convergent control flow.
So it should be checked that `__sycl_allocateLocalMemory` (or its source form - `group_local_memory`) does not occur:
- in a function (other than user lambda/functor)
- in a loop
- in a non-convergent control flow
Can we please add a TODO and file an issue?

I think it is OK to occur in convergent if. Some execution paths will just not use the global created for the allocation.

bader reviewed (commented):
@mlychkov, please, resolve merge conflicts.

kbobrovs commented on llvm/lib/Transforms/Utils/LowerWGLocalMemoryCall.cpp:
```diff
@@ -0,0 +1,114 @@
+//===-- LowerWGLocalMemoryCall.cpp - SYCL kernel local memory allocation --===//
```
Please move this to llvm/lib/SYCLLowerIR where SYCL-specific lowering passes congregate.

mlychkov replied:
Done

MrSidims commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,124 @@
+static bool lowerAllocaLocalMem(Module &M) {
+  SmallVector<CallInst *, 8> ToReplace;
+  for (Function &F : M) {
+    CallingConv::ID CC = F.getCallingConv();
```
```suggestion
    if (F.isDeclaration())
      continue;
    CallingConv::ID CC = F.getCallingConv();
```

mlychkov replied:
Thank you, added suggestion.

MrSidims commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,124 @@
+
+      // TODO: Static local memory allocation should be requested only in
+      // spir kernel scope.
+      assert((CC == llvm::CallingConv::SPIR_FUNC ||
```
I see that the logic here is to error out in case of finding SYCL_ALLOCLOCALMEM_CALL in a non-spir function, but still have you considered to move it above to have an early exit in case if the function isn't spir?

mlychkov replied:
Then there may be left calls of __sycl_allocateLocalMemory in some non spir function and compilation may fail later with error "unresolved __sycl_allocateLocalMemory", which is not clear for a user.

kbobrovs commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,125 @@
+static constexpr char LOCALMEMORY_GV_PREF[] = "WGLocalMem";
+
+namespace {
+class SYCLLowerWGLocalMemoryLegacy : public ModulePass {
```
Not clear why this is a `ModulePass` rather than `FunctionPass` - the pass does not seem to perform any cross-function actions. If none - suggest to make it FunctionPass

mlychkov replied:
I was guided by FunctionPass restrictions from LLVM documentation (https://llvm.org/docs/WritingAnLLVMPass.html#the-functionpass-class):
> 3. Add or remove global variables from the current Module.

kbobrovs commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,125 @@
+              CC == llvm::CallingConv::SPIR_KERNEL) &&
+             "WG static local memery can be allocated only in kernel scope");
+
+      ToReplace.push_back(CI);
```
Is there a reason to accumulate in `ToReplace` first rather than doing the replacement right here?

mlychkov replied:
Yes, the reason is that a kernel function body is changed (__sycl_allocateLocalMemory call instruction is removed). And instructions iterator is invalidated after this change.

kbobrovs commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,125 @@
+//
+// This pass replaces calls to __sycl_allocateLocalMemory(Size, Alignment)
+// function with allocation of memory in local address space at the kernel
+// scope.
```
Please do not duplicate the description. You can just redirect the reader to the header for it.

mlychkov replied:
Done

kbobrovs commented on llvm/include/llvm/SYCLLowerIR/LowerWGLocalMemory.h:
```diff
@@ -0,0 +1,33 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Replaces calls to __sycl_allocateLocalMemory(Size, Alignment) function with
```
The pass does more - adding new globals. This is also part of key actions the pass does, please describe shortly - what they model, that it is 1-per allocation, how IR changes.

mlychkov replied:
Thank you very much for a thorough review and helpful hints.
Modified pass description in the header.

kbobrovs reviewed (commented):
Thanks for moving. Few more comments.

Do you have end-to-end test, BTW or this is WIP? It is necessary for such functionality.

DenisBakhvalov commented on llvm/test/SYCLLowerIR/group_local_memory.ll:
```diff
@@ -0,0 +1,122 @@
+; RUN: opt -S -sycllowerwglocalmemory < %s | FileCheck --check-prefixes=CHECK,CHECK-CALL %s
```
I suggest cutting this test significantly. The only thing you care about is calls to `__sycl_allocateLocalMemory`, so the function under test can be as simple as having only one instruction. Then you can use `update_test_checks.py` for clarity, but it's up to you.

mlychkov replied:
Reduced.

DenisBakhvalov commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,125 @@
+      CallingConv::ID CC = F.getCallingConv();
+      assert((CC == llvm::CallingConv::SPIR_FUNC ||
+              CC == llvm::CallingConv::SPIR_KERNEL) &&
+             "WG static local memery can be allocated only in kernel scope");
```
memery -> memory

mlychkov replied:
Thank you for the catch. Fixed.

DenisBakhvalov commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -959,6 +960,10 @@ void EmitAssemblyHelper::EmitAssembly(BackendAction Action,
     PerModulePasses.add(createSPIRITTAnnotationsPass());
   }
 
+  // This pass should be always called for SYCL device code.
```
nit: don't you think that the comment has a low value?

mlychkov replied:
Added another comment.

mlychkov commented:
> Thanks for moving. Few more comments.
> 
> Do you have end-to-end test, BTW or this is WIP? It is necessary for such functionality.

End-to-end test for this feature is added by Semenov Sergey: https://github.com/intel/llvm-test-suite/pull/176.

AlexeySachkov commented on llvm/include/llvm/SYCLLowerIR/LowerWGLocalMemory.h:
```diff
@@ -0,0 +1,49 @@
+//===----------------------------------------------------------------------===//
+// This pass does the following for each allocate call to
+// __sycl_allocateLocalMemory(Size, Alignment) function at the kernel scope:
+// - inserts a global byte array of Size bytes with specified alignment
+//   in work group local address space.
```
What does `global byte array on local address space` mean? Is it a program-scope array or kernel-scope array?

UPD: looking at the code below, it seems to be a global (i.e. program-scope) pointer to a local memory - is it correct?

mlychkov replied:
Yes, that is right. I updated description.

AlexeySachkov commented on llvm/include/llvm/SYCLLowerIR/LowerWGLocalMemory.h:
```diff
@@ -0,0 +1,49 @@
+  PreservedAnalyses run(Module &M, ModuleAnalysisManager &);
+};
+
+ModulePass *createSYCLLowerWGLocalMemoryPass();
```
```suggestion
ModulePass *createSYCLLowerWGLocalMemoryPassLegacy();
```

mlychkov replied:
Updated.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,135 @@
+
+#define DEBUG_TYPE "LowerWGLocalMemory"
+
+static constexpr char SYCL_ALLOCLOCALMEM_CALL[] = "__sycl_allocateLocalMemory";
```
I assume that you are looking for calls of particular forward-declared functions and you are comparing names in the following manner:
```c++
      if (Callee->getName() != SYCL_ALLOCLOCALMEM_CALL)
        continue;
```

Please note that if `__sycl_allocateLocalMemory` is a forward-declared function somewhere in SYCL headers, then its name will be mangled (unless it is declared as `extern C`). Please check that with @sergey-semenov.

If the name is mangled, you should either use `.contains()` or (which is better) use `.startswith()` and include `_Z[some-number]` prefix into `SYCL_ALLOCLOCALMEM_CALL` reference string

mlychkov replied:
__sycl_allocateLocalMemory is declared as extern C function. So, there is no mangling and overload.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,135 @@
+  SmallVector<CallInst *, 8> ToReplace;
+  Function *allocaLocalMemF = nullptr;
+
+  for (Function &F : M) {
+    if (F.isDeclaration())
+      continue;
+
+    for (auto &I : instructions(F)) {
+      auto *CI = dyn_cast<CallInst>(&I);
+      Function *Callee = nullptr;
+      if (!CI || !(Callee = CI->getCalledFunction()))
+        continue;
+      if (Callee->getName() != SYCL_ALLOCLOCALMEM_CALL)
+        continue;
```
Instead of looking at each `call` instruction, you can find the definition of the pseudo-intrinsic you are looking for and traverse its users: I assume it would be more optimal solution. See intel/llvm#2610 - similar approach was applied to `SpecConstantsPass`

mlychkov replied:
Thank you for the hint. Improved implementation.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,135 @@
+      assert(Callee->isDeclaration() &&
+             "__sycl_allocateLocalMemory shouldn't have definition");
+
+      // TODO: Static local memory allocation should be requested only in
+      // spir kernel scope.
```
It is not clear to me what is the point of this `TODO` - is it for some other component or do we need some additional error handling (besides `assert`) here?

mlychkov replied:
Updated TODO description.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp:
```diff
@@ -0,0 +1,135 @@
+             "WG static local memory can be allocated only in kernel scope");
+
+      ToReplace.push_back(CI);
+      allocaLocalMemF = Callee;
```
Can this pseudo-intrinsic be overloaded?

mlychkov replied:
__sycl_allocateLocalMemory is declared as extern C function. So, there is no mangling and overload.

bader commented on llvm/lib/CodeGen/CMakeLists.txt:
```diff
@@ -211,6 +211,7 @@ add_llvm_component_library(LLVMCodeGen
   ProfileData
   Scalar
   Support
+  SYCLLowerIR
```
Why do we link `SYCLLowerIR` here?
According to my understanding it's not used in llvm's CodeGen.

mlychkov replied:
Sorry, I forgot to revert this chenge after implementation improvement. Removed.

bader commented on clang/test/CodeGenSYCL/group-local-memory.cpp:
```diff
@@ -0,0 +1,44 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -S -emit-llvm %s -o - | FileCheck --check-prefixes=CHECK,CHECK-UNOPT %s
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -S -emit-llvm %s -o - | FileCheck %s
```
I think llvm/test/SYCLLowerIR/group_local_memory.ll should be enough to validate the pass. Just make sure it covers both optimized and unoptimized IR.
Testing a single pass by running the whole compilation stack seems like an overkill to me.
Are there any other reasons to have this test in addition to llvm/test/SYCLLowerIR/group_local_memory.ll?

mlychkov replied:
The main perpose of current check is to test that pass is executed during device code compilation in clang in both cases when llvm passes are turned on and turned off.

bader replied:
There is much chipper way to achieve this with `-mllvm -debug-pass=Structure` - see https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenCUDA/link-device-bitcode.cu#L37-L41 as an example.
Test program might be an empty file, if I'm not mistaken.

mlychkov commented on clang/test/CodeGenSYCL/kernel-by-reference.cpp:
```diff
@@ -15,15 +15,15 @@ int simple_add(int i) {
   queue q;
 #if defined(SYCL2020)
-  // expected-warning@Inputs/sycl.hpp:301 {{Passing kernel functions by value is deprecated in SYCL 2020}}
+  // expected-warning@Inputs/sycl.hpp:312 {{Passing kernel functions by value is deprecated in SYCL 2020}}
```
Thank you. Improved.

bader reviewed (dismissed):
LGTM, although I'm not sure if it works for NVPTX target.

AlexeySachkov reviewed (approved):
LGTM, but the main LIT test for this change seems to be failing

premanandrao reviewed (approved):
FE changes look okay to me.

mlychkov commented:
@kbobrovs Could you please review last changes?

sergey-semenov commented:
> LGTM, although I'm not sure if it works for NVPTX target.

@mlychkov @bader The calling convention assertion https://github.com/intel/llvm/pull/3329/files#diff-a107684934906a250c3dc947338a6a685dc924e9246d6a33adc0b82f9b9de756R66-R68 fails for NVPTX (the default C convention is returned instead of the expected SPIR_FUNC/SPIR_KERNEL), but the E2E test passes after commenting the assertion out.

bader commented:
> > LGTM, although I'm not sure if it works for NVPTX target.
> 
> @mlychkov @bader The calling convention assertion https://github.com/intel/llvm/pull/3329/files#diff-a107684934906a250c3dc947338a6a685dc924e9246d6a33adc0b82f9b9de756R66-R68 fails for NVPTX (the default C convention is returned instead of the expected SPIR_FUNC/SPIR_KERNEL), but the E2E test passes after commenting the assertion out.

I suggest removing the assert then.

bader commented:
@kbobrovs, ping.

kbobrovs commented:
> @kbobrovs, ping.

@bader, @mlychkov - LGTM, except the diagnostics - I'd request at least TODO + issue (see my comment above).

kbobrovs reviewed (approved):
Approved. As agreed with @mlychkov, TODO + issue will follow.
